python/lsst/summit/utils/tmaUtils.py

```diff
@@ -0,0 +1,1588 @@
+                # first row of the next event, because you were in that state
+                # right up until that state change. However, if that event is
+                # a) the last one of the day and b) runs right up until the end
+                # of the dataframe, then there isn't another row, so this will
+                # overrun the array.
+                #
+                # If the data is for the current day then this isn't a worry,
+                # as we're likely still taking data, and this event will likely
+                # close yet, so we don't issue a warning, and simply drop the
+                # event from the list.
+
+                # However, if the data is for a past day then no new data will
+                # come to close the event, so allow the event to be "open", and
+                # issue a warning
+                if dataIsForCurrentDay:
+                    self.log.info("Discarding open (likely in-progess) final event from current day's events")
+                    parsedStates = parsedStates[:-1]
+                else:
+                    self.log.warning("Last event ends open, forcing it to end at end of the day's data")
+                    # it's a tuple, so (deliberately) awkward to modify
+                    parsedStates[-1] = (lastEvent[0], lastEvent[1] - 1, lastEvent[2], lastEvent[3])
+
+        return parsedStates
+
+    def addBlockDataToEvents(self, events):
+        """Find all the block data in the EFD for the specified events.
+
+        Finds all the block data in the EFD relating to the events, parses it,
+        from the rows of the dataframe, and adds it to the events in place.
+
+        Parameters
+        ----------
+        events : `lsst.summit.utils.tmaUtils.TMAEvent` or
+                 `list` of `lsst.summit.utils.tmaUtils.TMAEvent`
+            One or more events to get the block data for.
+        """
+        events = ensure_iterable(events)
+        events = sorted(events)
+
+        # Get all the data in one go and then clip to the events in the loop.
+        # This is orders of magnitude faster than querying individually.
+        allData = getEfdData(self.client,
+                             "lsst.sal.Script.logevent_state",
+                             begin=events[0].begin,  # time ordered, so this is the start of the window
+                             end=events[-1].end,  # and this is the end
+                             noWarn=True)
+        if allData.empty:
+            self.log.info('No block data found for the specified events')
+            return {}
+
+        blockPattern = r"BLOCK-(\d+)"
+        blockIdPattern = r"BL\d+(?:_\w+)+"
+        sitcomPattern = r"SITCOM-(\d+)"
+
+        for event in events:
+            eventData = clipDataToEvent(allData, event)
+            if eventData.empty:
+                continue
+
+            blockNums = set()
+            blockIds = set()
```

Comment:
I see what you're saying, but I think that will be quite a bit slower because of all the checks. More importantly though, all this code is going to be removed/redone on DM-39852 as soon as this is merged and I'm back from vacation, so this will be moot very soon indeed.